### ISSUE TITLE
Configure sliding expiration of authentication cookie

### ISSUE THREAD

**Issue by thomasfrisko** [IdentityServer/IdentityServer4#4714]
We use both .NET Core Identity and IdentityServer4 and want to implement 'Remember me' functionality. For this we decided to use cookies (as shown in examples). When the user does not check the box 'Remember me', the authentication cookie will be a session cookie, otherwise if the user does check the box, they will get a persistent cookie. 
We use .NET Core Identity as "DefaultAuthenticateScheme" with configuration of the application cookie as follows:

```
services.ConfigureApplicationCookie(o =>
{
   o.ExpireTimeSpan = TimeSpan.FromMinutes(30);
   o.SlidingExpiration = true;
});
```

The expiration time of the cookie is set correctly, however, the sliding expiration does not seem to work. 
If understood correctly, if we attempt to login (call the authorize endpoint) after 15 minutes (when expiration is 30), the cookie should be recreated with a new expiration lifetime so it lasts for another 30 minutes, however this never happens. This means the cookie will only work for 30 minutes, then the user will have to enter credentials again with these settings.

To sum up, I would like to know why setting `o.SlidingExpiration = true`(EDITED) does not recreate the cookie with a new expiration when requesting the server while the cookie is more than halfway through its expiration window.

Thanks in advance!

**Comment by brockallen** [comment]
`o.Authentication.CookieSlidingExpiration` is configuring the settings on the cookie authentication handler that IdentityServer automatically registers, but if you're using ASP.NET Identity then you're likely using theirs and not the one from IdentityServer.

**Comment by thomasfrisko** [comment]
Sorry, I did not mean `o.Authentication.CookieSlidingExpiration = true`, i meant setting `o.SlidingExpiration = true` using the Identity's `services.ConfigureApplicationCookie()`. (Im editing this in the original post)

When requesting the authorize endpoint, the `DefaultUserSession` service is used to authenticate the user. This uses an extension method on the `HttpContext` called `GetCookieAuthenticationSchemeAsync`. This method looks for the `CookieAuthenticationScheme`. if this scheme has not been set through `IdentityServerOptions`, it will default to `DefaultAuthenticateScheme`. In this case `DefaultAuthenticateScheme` is set by Identity to the `ApplicationScheme` ('Identity.Application'). As seen in the image below, this seems to be set correctly. 
![auth-scheme](https://user-images.githubusercontent.com/7311868/89505164-7c2ddc00-d7c9-11ea-8b35-6164ed9380ff.png)

The handler for this cookie also seem to have the options correctly set.
![auth-options](https://user-images.githubusercontent.com/7311868/89505248-9d8ec800-d7c9-11ea-817f-d509693a1b15.png)


The `ApplicationScheme` can be configured by the method `ConfigureApplicationCookie`, therefore I assumed setting both `ExpirationTimeSpan = TimeSpan.FromMinutes(30)` and `SlidingExpiration=true` would work. However, only `ExpirationTimeSpan` seem to have an effect on the authentication cookie (where it correctly sets the expiration as seen in the image below).
![Identity](https://user-images.githubusercontent.com/7311868/89503821-64edef00-d7c7-11ea-99f6-8f7600ac5703.PNG)

**Comment by brockallen** [comment]
So given this update, you're basically having issues with the cookie from ASP.NET Identity... so it's nothing in IdentityServer that's affecting the behavior you get.

**Comment by thomasfrisko** [comment]
Yes it is the cookie from ASP .NET Identity I am having issues with, however, I do not know whether not updating its expiration time has something to do with IdentityServer or if it is completely unrelated.

I am logging in from an angular client application using oidc-client javascript library. As mentioned previously the cookie expiration is set initially but never updated afterwards with the sliding expiration setting. 

I managed to make it update the cookie expiration slide by accessing the IdentityServer directly by opening a new browser tab and then enter the url instead of using the angular client to call authorize endpoint after initial login (login with angular client -> open new browser tab -> enter url for identityServer webservice), so the setting `SlidingExpiration=true` works, however, it will not trigger through the angular client (maybe related to how javascript does http requests?).

I found this post on SO https://stackoverflow.com/questions/53292975/identityserver-session-cookie-is-not-sliding 
where it is discussed how to configure the sliding expiration behavior for IdentityServer session cookie. I used the code in the accepted answer (modified it a bit to make it work with identity authentication cookie): 
`o.Events.OnValidatePrincipal = context =>
{
   string EXPIRES_KEY = ".expires";
   string UTC_DATE_TIME_FORMAT = "r";

    if (context.Principal.Identity.IsAuthenticated)
    {
       if (context.Properties.Items.ContainsKey(EXPIRES_KEY) && 
          context.Request.Path.Value.StartsWith("/connect/authorize"))
       {
           var expiresAt = DateTimeOffset.Parse(context.Properties.Items[EXPIRES_KEY]);
           if (DateTimeOffset.UtcNow <= expiresAt)
           {
               context.ShouldRenew = true;
               context.Properties.Items[EXPIRES_KEY] =
               DateTimeOffset.UtcNow.AddMinutes(30)
               .ToString(UTC_DATE_TIME_FORMAT, CultureInfo.InvariantCulture);
            }
         }
    }
    return Task.CompletedTask;
};`

This code extends the expiration time of the authentication cookie (also when using the angular client), so I will use this approach since it solves my issue, however, it seems strange that `SlidingExpiration=true` does not work when requesting the authorize endpoint through the angular client since IdentityServer clearly uses the cookie from Identity to ensure the user is authenticated. (Maybe I'm missing some understanding of how the `CookieAuthenticationHandler` works)

**Comment by brockallen** [comment]
Not sure -- you should debug the HTTP traffic to diagnose.

**Comment by Dommicentl** [comment]
Maybe this is also caused by #4812
